### PR TITLE
Enter the JobDescriptor.  Closes #496

### LINKED ISSUE #496: Create CallDescriptor abstraction

### PR DESCRIPTION
**Pull request by mcovarr**
This introduces a ~~`CallDescriptor`~~ `JobDescriptor` abstraction that I hope resembles what we'd want in the world of pluggable backends.  Although this is a backend class, it holds a reference to a `WorkflowDescriptor` which still has tons of references to engine things which we plan to refactor out, possibly into backend.  `CallDescriptor` currently interacts with `BackendCall`s because that's what Cromwell has right now.  :smile: 

There was one very special bit of weirdness about our backend handling exposed by `RetryableCallsSpec` where the backend passed to `WorkflowManagerActor` is not necessarily the same backend used to build `WorkflowDescriptor`s. :frowning:  That could use some extra scrutiny by reviewers.

### CODE REVIEW

**Comment by mcovarr** [comment]
Reviewers: @Horneth and @gauravs90

File: src/main/scala/cromwell/engine/backend/BackendCallDescriptor.scala [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import wdl4s.values.WdlValue
+
+
+sealed trait CallDescriptor[K <: CallKey] {
+  def workflowDescriptor: WorkflowDescriptor
+  def key: K
+  def locallyQualifiedInputs: CallInputs
+}
+
+final case class BackendCallDescriptor(workflowDescriptor: WorkflowDescriptor,
```
Comment (jainh) [inline src/main/scala/cromwell/engine/backend/BackendCallDescriptor.scala]: Why do we need workflowDescriptor as a parameter when it is extending a trait that contains the information already? or do we need to defining methods for as getter in trait?

Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/BackendCallDescriptor.scala]: The trait says that an implementer must have a workflowDescriptor. This implementation says that it is an argument to the constructor.

Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/BackendCallDescriptor.scala]: Also note that this demonstrates how when implementing a scala trait you can always refine downward (def -> lazy val -> val -> var) for your concrete instantiation as the implementation is using a val an the trait specified a def

File: src/main/scala/cromwell/engine/callactor/CallActor.scala [hunk 2]
```diff
@@ -74,15 +74,16 @@ object CallActor {
...
+  def props(backendCallDescriptor: BackendCallDescriptor): Props = Props(new BackendCallActor(backendCallDescriptor))
+  def props(finalCallDescriptor: FinalCallDescriptor) = Props(new FinalCallActor(finalCallDescriptor))
 }
 
 /** Actor to manage the execution of a single call. */
-trait CallActor extends LoggingFSM[CallActorState, CallActorData] with CromwellActor {
-  def key: CallKey
-  def workflowDescriptor: WorkflowDescriptor
+// FIXME It feels like I shouldn't need to restate the type bounds of CallDescriptor's CallKey type variable.
+trait CallActor[D <: CallDescriptor[_ <: CallKey]] extends LoggingFSM[CallActorState, CallActorData] with CromwellActor {
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/callactor/CallActor.scala]: someone immersed in higher kindedness could probably straighten this out @geoffjentry

File: src/main/scala/cromwell/engine/backend/BackendCallDescriptor.scala (hunk 1 above)
Comment (jainh) [inline src/main/scala/cromwell/engine/backend/BackendCallDescriptor.scala]: Seems like kind of access methods for the properties/arguments to case class. As long as they are read only operations and don't break immutability looks fine.

**Comment by Horneth** [comment]
:+1:

File: src/main/scala/cromwell/engine/backend/BackendCallDescriptor.scala (hunk 1 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/BackendCallDescriptor.scala]: They're properties of the case class, by definition they're both read only and immutable.

File: src/main/scala/cromwell/engine/callactor/CallActor.scala (hunk 2 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/callactor/CallActor.scala]: Perhaps, but not at the moment. If you still care later on I'll take a look

Comment (mcovarr) [inline src/main/scala/cromwell/engine/callactor/CallActor.scala]: no rush, I don't think it's going anywhere :stuck_out_tongue:

File: src/main/scala/cromwell/engine/backend/BackendCall.scala [hunk 3]
```diff
@@ -84,11 +83,24 @@ trait BackendCall {
...
+    * }
+    *
+    * Then locallyQualifiedInputs could be Map("some_dir" -> WdlFile("/some/path"))
+    */
+  def locallyQualifiedInputs: CallInputs = callDescriptor.locallyQualifiedInputs
+
+  /**
    * Backend which will be used to execute the Call
    */
   def backend: Backend
```
Comment (gauravs90) [inline src/main/scala/cromwell/engine/backend/BackendCall.scala]: Perhaps, this can also be derived from the `CallDescriptor.workflowDescriptor` ?

File: src/main/scala/cromwell/engine/backend/BackendCall.scala [hunk 4]
```diff
@@ -105,18 +117,6 @@ trait BackendCall {
...
-    * }
-    *
-    * Then locallyQualifiedInputs could be Map("some_dir" -> WdlFile("/some/path"))
-    */
-  def locallyQualifiedInputs: CallInputs
-
-  /**
    * Implementation of the WDL Standard Library Functions, used to evaluate WdlExpressions
    */
   def engineFunctions: WdlStandardLibraryFunctions
```
Comment (gauravs90) [inline src/main/scala/cromwell/engine/backend/BackendCall.scala]: This too?

**Comment by gauravs90** [comment]
*Big picture:* This PR introduces a `CallDescriptor` wrapper around some other stuff.  How do you envision this to be of use in PBE? Do you see this as turning into a `TaskDescriptor`? Does it reduce any effort later on?

**Comment by mcovarr** [comment]
`TaskDescriptor` was slightly misnamed as it described a call and not a task (many calls can have the same underlying task).  So this is meant to be the new `TaskDescriptor` and should be able to describe a call in generic terms such that it could be executed on any backend implementation.

File: src/main/scala/cromwell/engine/backend/BackendCall.scala (hunk 3 above)
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/BackendCall.scala]: Looks like it, let me try.

File: src/main/scala/cromwell/engine/backend/BackendCall.scala (hunk 4 above)
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/BackendCall.scala]: I thought that might work but this reference is actually to `CallEngineFunctions` and not the `WorkflowEngineFunctions` that are in the `WorkflowDescriptor`.  I'll change the type here to be more explicit.

File: src/main/scala/cromwell/engine/backend/BackendCall.scala (hunk 3 above)
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/BackendCall.scala]: That unfortunately doesn't work due to the very tight coupling between `Backend` and `BackendCall`.  The subtypes of `BackendCall` refine the `Backend` to match (e.g. `JesBackendCall` redefines `backend` to be a `JesBackend`) and make references to methods in the subtype backends that are not present in the base `Backend` type.

**Comment by mcovarr** [comment]
@gauravs90 if this looks ok to you could you please give a :+1:?  Thanks

**Comment by mcovarr** [comment]
tl;dr I'd like to squash / rebase / merge this despite a test failure during one run since I think that failure was due to unrelated Docker pull issues.



So one build for this branch failed:

https://travis-ci.org/broadinstitute/cromwell/builds/113532462

The first failure was a docker test, and looking at this more closely something seems to have gone awry pulling the Docker image.  Our build scripts should pre-pull `ubuntu:latest` and normally this takes about 10 seconds and produces a nice success message.  In this run the Docker image pull took more than 43 seconds and the success message appears to be cut off:

```
Pulling repository docker.io/library/ubuntu
age for ubuntu:latest
``` 

The Docker test looks like it's going fine until it's time to actually run a call, at which point there are no log messages for 16 seconds, and when the log message does arrive it seems to indicate a timeout:

```
[INFO] [03/03/2016 23:43:02.128] [test-system-akka.actor.default-dispatcher-2] [WorkflowActor(akka://test-system)] WorkflowActor [UUID(299b2fc4)]: persisting status of a to Starting.
[WARN] [03/03/2016 23:43:18.664] [test-system-akka.actor.default-dispatcher-4] [akka://test-system/system/IO-TCP/selectors/$a/1] received dead letter from Actor[akka://test-system/user/IO-HTTP/group-0/1#-1001288108]: Write(ByteString(),spray.io.SslTlsSupport$WriteChunkAck$@22a4ed01)
```

There's another 13 second hang shortly thereafter:

```
[INFO] [03/03/2016 23:43:19.002] [test-system-akka.actor.default-dispatcher-10] [WorkflowActor(akka://test-system)] WorkflowActor [UUID(299b2fc4)]: persisting status of a to Running.
[INFO] [03/03/2016 23:43:32.134] [pool-7-thread-13-ScalaTest-running-CallCachingWorkflowSpec] [akka://test-system/user/$$h] WorkflowManagerActor submitWorkflow input id = None, effective id = c21e652b-b5f0-4435-a390-b1d61d1c9b4a
```

Next it looks like a test is started up while pointing to the same in-memory db as this paused workflow.  The paused workflow is interpreted as a workflow needing restart, so another concurrent copy is launched.  

```
[INFO] [03/03/2016 23:43:32.139] [ForkJoinPool-3-worker-3] [akka://test-system/user/$$h] WorkflowManagerActor Found 1 workflow to restart.
[INFO] [03/03/2016 23:43:32.139] [ForkJoinPool-3-worker-3] [akka://test-system/user/$$h] WorkflowManagerActor Restarting workflow ID: 299b2fc4-6a26-462f-96e3-1281f172d197
[INFO] [03/03/2016 23:43:32.152] [ForkJoinPool-3-worker-3] [akka://test-system/user/$$h] Invoking restartableWorkflow on 299b2fc4
[INFO] [03/03/2016 23:43:32.153] [ForkJoinPool-3-worker-3] [akka://test-system/user/$$h] WorkflowManagerActor submitWorkflow input id = Some(299b2fc4-6a26-462f-96e3-1281f172d197), effective id = 299b2fc4-6a26-462f-96e3-1281f172d197
```

This quickly falls afoul of a uniqueness constraint:

```
[ERROR] [03/03/2016 23:43:32.236] [ForkJoinPool-3-worker-1] [WorkflowActor(akka://test-system)] WorkflowActor [UUID(299b2fc4)]: Could not persist runtime attributes
java.sql.SQLIntegrityConstraintViolationException: integrity constraint violation: unique constraint or index violation; UK_RUNTIME_ATTRIBUTE table: RUNTIME_ATTRIBUTES
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCPreparedStatement.fetchResult(Unknown Source)
	at org.hsqldb.jdbc.JDBCPreparedStatement.executeUpdate(Unknown Source)
```

From that point on it's basically a trainwreck of tests cross-talking.

File: src/main/scala/cromwell/engine/backend/JobDescriptor.scala [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+final case class BackendCallJobDescriptor(workflowDescriptor: WorkflowDescriptor,
+                                          key: BackendCallKey,
+                                          locallyQualifiedInputs: CallInputs = Map.empty) extends JobDescriptor[BackendCallKey] {
+
+  // PBE temporarily still required.  Once we have call-scoped Backend actors they will know themselves and the
+  // backend won't need to be in the WorkflowDescriptor and this method won't need to exist.
+  def backend = workflowDescriptor.backend
+}
```
Comment (scottfrazer) [inline src/main/scala/cromwell/engine/backend/JobDescriptor.scala]: I'm not really a big fan of the name `BackendCallJobDescriptor`... maybe just `CallJobDescriptor`?

Comment (gauravs90) [inline src/main/scala/cromwell/engine/backend/JobDescriptor.scala]: I agree.. Too many adjectives :laughing:

Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/JobDescriptor.scala]: This nomenclature is to disambiguate between FinalCalls and is used pretty consistently everywhere else.  On the plus side the need for the disambiguation will hopefully go away once FinalCalls run on the backend like BackendCalls do now.

Comment (gauravs90) [inline src/main/scala/cromwell/engine/backend/JobDescriptor.scala]: Yes yes.. It was just a suggestion.This will anyways will be refactored later on :smile:

**Comment by gauravs90** [comment]
Aye :+1:
